cuda_core/tests/conftest.py

```diff
@@ -1,36 +1,56 @@
-# Copyright 2024 NVIDIA Corporation.  All rights reserved.
-#
-# Please refer to the NVIDIA end user license agreement (EULA) associated
-# with this source code for terms and conditions that govern your use of
-# this software. Any use, reproduction, disclosure, or distribution of
-# this software and related documentation outside the terms of the EULA
-# is strictly prohibited.
-try:
-    from cuda.bindings import driver
-except ImportError:
-    from cuda import cuda as driver
-
-import pytest
-
-from cuda.core.experimental import Device, _device
-from cuda.core.experimental._utils import handle_return
-
-
-@pytest.fixture(scope="function")
-def init_cuda():
-    device = Device()
-    device.set_current()
-    yield
-    _device_unset_current()
-
-
-def _device_unset_current():
-    handle_return(driver.cuCtxPopCurrent())
-    with _device._tls_lock:
-        del _device._tls.devices
-
-
-@pytest.fixture(scope="function")
-def deinit_cuda():
-    yield
-    _device_unset_current()
+# Copyright 2024 NVIDIA Corporation.  All rights reserved.
+#
+# Please refer to the NVIDIA end user license agreement (EULA) associated
+# with this source code for terms and conditions that govern your use of
+# this software. Any use, reproduction, disclosure, or distribution of
+# this software and related documentation outside the terms of the EULA
+# is strictly prohibited.
+
+import glob
+import os
+import sys
+
+try:
+    from cuda.bindings import driver
+except ImportError:
+    from cuda import cuda as driver
+
+import pytest
+
+from cuda.core.experimental import Device, _device
+from cuda.core.experimental._utils import handle_return
+
+
+@pytest.fixture(scope="function")
+def init_cuda():
+    device = Device()
+    device.set_current()
+    yield
+    _device_unset_current()
+
+
+def _device_unset_current():
+    handle_return(driver.cuCtxPopCurrent())
+    with _device._tls_lock:
+        del _device._tls.devices
+
+
+@pytest.fixture(scope="function")
+def deinit_cuda():
+    yield
+    _device_unset_current()
+
+
+# samples relying on cffi could fail as the modules cannot be imported
+sys.path.append(os.getcwd())
+
+
+@pytest.fixture(scope="session", autouse=True)
+def clean_up_cffi_files():
+    yield
+    files = glob.glob(os.path.join(os.getcwd(), "_cpu_obj*"))
+    for f in files:
+        try:  # noqa: SIM105
+            os.remove(f)
+        except FileNotFoundError:
```

Comment:
I would like to do the opposite: Ensure the artifacts do not exist after testing (regardless if tests succeed or not). The reason is that: I don't want  any artifact to remain after the tests finish. I find it troublesome having to update `.gitignore` to skip the artifacts (and, depending on where we run the tests, the artifact location could change). Without an RAII-like clean up, they'd still remain on the file system after tests, and a subsequent local run (outside of pytest) could accidentally reuse them, which is considered a side effect that I do not like.

Reply:
Sounds good.

I'm still wondering when/why the `except FileNotFoundError` is needed, but it most likely will not do any harm to have it here.

Reply:
It could happen if CFFI compilation fails for whatever reason (so that there's no artifact to remove).

Reply:
The list of `files` is the result of `glob.glob(os.path.join(os.getcwd(), "_cpu_obj*"))`, which will only produce what actually exists in the filesystem (it could be empty for example).

```
>>> import glob
>>> glob.glob("_cpu_obj*")
[]
```

This could only go wrong if something concurrently deletes the files. (I wouldn't want to mask that.)

Reply:
> This could only go wrong if something concurrently deletes the files. (I wouldn't want to mask that.)

You might be onto something. I [wrote this snippet](https://github.com/NVIDIA/cuQuantum/blame/b3e6d623ecb7d95621660d64f864e09c93c22028/python/tests/cuquantum_tests/__init__.py#L28-L34) a while back and perhaps back then the tests there were run under [pytest-xdist](https://github.com/NVIDIA/cuQuantum/blob/b3e6d623ecb7d95621660d64f864e09c93c22028/python/tests/requirements.txt#L2) which parallelized the tests.

Reply:
FWIW, I asked chatgpt for help with temp files. The answer suggests that it's really easy to avoid filesystem races with pytest:

https://docs.google.com/document/d/1IuhmAgtITcnvrEJY5BTkCb-taWw2Soj2YBmU535Ttoc/edit?usp=sharing

Use `tmp_path` Fixture:

`tmp_path` provides a Path object from Python's pathlib, which is modern and more feature-rich.
This is the recommended approach for most use cases as Path is more intuitive and robust.